KEDamage.cs

```diff
@@ -8,58 +8,73 @@ class KEDamage:Damage
+            /// </summary>
             public float FrictionFactor;
         }
 
-        private KineticData data;
-        private Damage.Target target;
+        private KineticData _keData;
 
-        public KEDamage(KineticData d,Damage.Target t):base(EDamageTypes.KE)
+        public KEDamage(KineticData d,Damage.Target t):base(EDamageTypes.KE, t)
         {
-            data = d;
-            target = t;
+            _keData = d;
         }
 
-        override public float DealDamage()
+        public override Target CalculateDamage()
         {
+            Target finalState = _target;
+
             // Calculate attenuation of air friction
-            data.Pierce = CalculateKEAttenuationSimple(data.Pierce,data.Distance,data.FrictionFactor);
+            _keData.Pierce = CalculateKEAttenuationSimple(_keData.Pierce,_keData.Distance,_keData.FrictionFactor);
 
             // Calculate effects of ERA
-            data.Pierce = CalculatePostERAPierce(data.Pierce,target.ERAData.KEFractionMultiplier);
+            float finalERA = Math.Max(0.0f, _target.ERAData.CurrentValue
+             - _keData.Pierce*_target.ERAData.KEFractionMultiplier);
+            finalState.ERAData.CurrentValue = finalERA;
+            
+            _keData.Pierce = CalculatePostERAPierce(_keData.Pierce,_target.ERAData.KEFractionMultiplier);
+
+            // Armor degradation
+            float finalArmor = Math.Max(0.0f,
+             _target.Armor - (_keData.Pierce / _target.Armor)*_keData.ArmorDegradationFactor);
+            finalState.Armor = finalArmor;
 
             // Calculate final damage
-            float finalDamage = Math.Max(0.0f, target.Armor - data.Pierce)*data.HealthKEFactor;
-            return finalDamage;
+            float finalDamage = Math.Max(0.0f, _keData.Pierce - _target.Armor)*_keData.HealthDamageFactor;
+            float finalHealth = Math.Max(0.0f, _target.Health - finalDamage);
+            finalState.Health = finalHealth;
+            
+            return finalState;
         }
 
-        // Then we implement a separate function for armour degradation,
-        // use callback function to update target armour value in its data model class
-
-        #region DamageCalculation
 
         private static float CalculateKEAttenuationSimple(float pierce, float distance, float frictionFactor)
         {
             float finalPierce =  Math.Exp(-frictionFactor*distance)*pierce;
             return finalPierce;
         }
         
-        private static float CalculatePostERAPierce(float pierce, float eraFrictionMultiplier)
+        private static float CalculatePostERAPierce(float pierce, float eraFractionMultiplier)
         {
-            float finalPierce = pierce*(1-eraFrictionMultiplier);
-
-            // Use some callback function to update target's era value here
-
+            float finalPierce = pierce*(1-eraFractionMultiplier);
             return finalPierce;
```

Comment:
You could just return `pierce * ( 1 - eraFractionMultiplier);` because the method name states what it's getting. Up to you.